File: Binary_tree_general/Tree_traversal/BFT_DFT.py
class TreeNode():
    def __init__(self, x: int):
        self.val = x
        self.left = None
        self.right = None

'''
BFT : Breadth First Traversal


DFT : Depth First Traversal
Preorder : visit a node before traversing : file directory listing
Inorder : Traverse -> visit in the middle : convert a binary search tree to a sorted list
Postorder : visit a node after traversing : file size calculation
'''

class Tree():

    def __init__(self, root: TreeNode = None):
        self.root = root

    def visit(self, node : TreeNode):
        print(node.val)
        '''
        사실 visit  할 때, 찾고자 하는 node인지 여부를 확인하면
        BFT-> BFS(search)
        DFT-> DFS(search)
        가 되는 것. 찾아보니 그냥 동일한 개념이랜다.
        '''
    
    #BFT
    def BFT(self):
        if self.root == None:
            return
        q = [self.root] #or q = deque([self.root])

        while q: # empty 여부 조건 검사 할 때가 한 level 임.(depth)
            curNode = q.pop(0)

            self.visit(curNode) #
            if curNode.left is not None:
                q.append(curNode.left)
            if curNode.right is not None:
                q.append(curNode.right)


    #DFT
    ## preorder
    def __DFT_preorderHelp(self, curNode):
        if curNode == None: ##  curNode가 leaf 이상으로 넘어갔을 경우, 함수 종료.
            return
        
        #먼저 방문하고, 다음 child 로 넘어가기. child 먼저.
        self.visit(curNode)
        self.__DFT_preorderHelp(curNode.left)
        self.__DFT_preorderHelp(curNode.right)

    def DFT_preorder(self):
        self.__DFT_preorderHelp(self.root) # root부터 시작

    
    ##inorder
    def __DFT_inorderHelp(self, curNode):
        if curNode == None:
            return
        
        self.__DFT_inorderHelp(curNode.left)
        self.visit(curNode)
        self.__DFT_inorderHelp(curNode.right)

    def DFT_inorder(self):
        self.__DFT_inorderHelp(self.root)
    
    ##post order
    def __DFT_postorderHelp(self, curNode):
        if curNode == None:
            return
        
        self.__DFT_postorderHelp(curNode.left)
        self.__DFT_postorderHelp(curNode.right)
        self.visit(curNode)

    def DFT_postorder(self):
        self.__DFT_postorderHelp(self.root)


def main():
    node1 = TreeNode(1)
    node2 = TreeNode(2)
    node3 = TreeNode(3)
    node4 = TreeNode(4)
    node5 = TreeNode(5)


    node1.left = node2
    node1.right = node3
    node2.left = node4
    node2.right = node5


    tree = Tree(node1)

    print("BFT")
    tree.BFT()

    print("DFT")

    print("preorder")
    tree.DFT_preorder()
    print("inorder")
    tree.DFT_inorder()
    print("postorder")
    tree.DFT_postorder()


if __name__ == "__main__":
    main()

File: Binary_tree_general/Tree_traversal/BFT_DFT.cpp
#include <iostream>
#include <queue>
using namespace std;

class TreeNode{
    public:
        int val;
        TreeNode* left;
        TreeNode* right;
        TreeNode(int x) : val(x), left(nullptr), right(nullptr){

        }
        ~TreeNode(){
            delete left;
            delete right;
        }
};


class Tree{
    private:
        TreeNode* root;

    public:
        Tree(TreeNode * root=nullptr): root(root){

        }
        ~Tree(){
            delete root;//이렇게 하면 root 부터 child 모두 재귀적으로 삭제
        }

        void visit(TreeNode * node){
            cout<<node->val<<" ";
        }


        void BFT(){
            if (root == nullptr){
                return;
            }

            queue<TreeNode*> q;
            q.push(root);

            while(!q.empty()){
                TreeNode* curNode = q.front();
                q.pop();

                visit(curNode);
                if (curNode->left !=nullptr){
                    q.push(curNode->left);
                }
                if (curNode->right!=nullptr){
                    q.push(curNode->right);
                }
            }
        }

        void DFT_preorderhelp(TreeNode* curNode){
            if (curNode == nullptr){
                return;
            }
            visit(curNode);
            DFT_preorderhelp(curNode->left);
            DFT_preorderhelp(curNode->right);
        }

        void DFT_preorder(){
            DFT_preorderhelp(root);
        }
        
        void DFT_inorderhelp(TreeNode* curNode){
            if (curNode == nullptr){
                return;
            }
            DFT_inorderhelp(curNode->left);
            visit(curNode);
            DFT_inorderhelp(curNode->right);
            
        }

        void DFT_inorder(){
            DFT_inorderhelp(root);
        }

        void DFT_postorderhelp(TreeNode* curNode){
            if (curNode == nullptr){
                return;
            }
            DFT_postorderhelp(curNode->left);
            DFT_postorderhelp(curNode->right);
            visit(curNode);
            
        }

        void DFT_postorder(){
            DFT_postorderhelp(root);
        }
};


int main(){

    TreeNode* node1 = new TreeNode(1);
    TreeNode* node2 = new TreeNode(2);
    TreeNode* node3 = new TreeNode(3);
    TreeNode* node4 = new TreeNode(4);
    TreeNode* node5 = new TreeNode(5);
    
    node1->left = node2;
    node1->right = node3;
    node2->left = node4;
    node2->right = node5;

    Tree tree(node1);

    cout<<"BFT"<<endl;
    tree.BFT();
    cout<<endl;
    cout<<"preorder"<<endl;
    tree.DFT_preorder();
    cout<<endl;
    cout<<"inorder"<<endl;
    tree.DFT_inorder();
    cout<<endl;
    cout<<"postorder"<<endl;
    tree.DFT_postorder();


     //이렇게 하면 root 부터 child 모두 재귀적으로 삭제
     //delete tree를 안해도 되는게, tree는 동적할당이 아니다. 
     //main을 벗어나면 tree가 destructor 가 발동되는데, 그때, root가 delete되면서 자식노드들 모두 free됨.

    return -1;
}